add basic ldif reader

vetinari opened the pull request.

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+// A basic LDIF parser. This one does currently just support LDIFs like
+// they're generated by tools like ldapsearch(1) / slapcat(8). Change
+// records are not supported.
+type LDIF struct {
```
mixing the parser and the data is unusual... I would expect an LDIFParser that decodes a []byte into an LDIF struct

liggitt replied:
maybe something along the lines of https://golang.org/pkg/encoding/json/#Unmarshal, e.g. `ldif.Unmarshal(io.Reader, *ldif.LDIF) error`

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+type LDIF struct {
+	RelaxedParser bool // ignore parsing errors in a line
+	Entries       []*Entry
+	curLine       int
```
use of `curLine`,`version`, and `changeType` seems local to Parse... do they need to be member fields?

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+	changeType    string
+}
+
+var CR byte = '\x0D'
```
make these unexported consts

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+var SPACE byte = ' '
+var SPACES = string(SPACE)
+
+func (l *LDIF) newError(msg string) error {
```
a structured error would be better so the line number, etc is retrievable
```
type LDIFParseError struct {
  Line int
  Message string
}
func (e LDIFParseError) Error() string {
  ...
}
```

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+				switch nextLine[0] {
+				case Comment:
+					continue
+				case SPACE:
```
I'm not familiar enough with the ldif schema to know... is a single leading space enough to determine this?

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+	// for i, line := range lines {
+	//     fmt.Fprintf(os.Stderr, "% 2d %s\n", i, line)
+	// }
+	if l.version == 0 && bytes.HasPrefix(lines[0], []byte("version:")) {
```
again... I don't know enough about LDIF to know... is `version` once per file or per entry? if per file, it doesn't seem like this belongs in parseEntry

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+	dn := val
+
+	lines = lines[1:]
+	if bytes.HasPrefix(lines[0], []byte("changetype:")) {
```
seems like there are several chances to panic with accesses to `lines[0]` if too few lines are passed in

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+			return errors.New("OID with at least 2 consecutive dots")
+		case c == '.':
+			lastDot = true
+		case c >= 0x30 && c <= 0x39:
```
does comparing to runes work (`c >= '0' && c <= '9'`)? if so, it would be more readable

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+		return errors.New("empty attribute name")
+	}
+	switch {
+	case attr[0] >= 0x41 && attr[0] <= 0x5A:
```
same here, I think you can actually compare to the runes for readability

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+package ldap
```
I think this should probably go in an `ldif` subpackage

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+
+	for {
+		l.curLine++
+		nextLine, err = reader.ReadBytes(LF)
```
if I'm reading the ldif rfc correctly, all the lines can be handled as strings. treating lines as strings throughout might clean up a lot of []byte casts

liggitt commented on ldif_test.go:
```diff
@@ -0,0 +1,118 @@
+telephonenumber: +1 408 555 1212
+`
+
+func TestLDIFParseRFC2849Example(t *testing.T) {
```
would be good to validate the resulting entries, in addition to making sure the parse didn't error

liggitt commented on ldif_test.go:
```diff
@@ -0,0 +1,118 @@
+cn: Bjorn Jensen
+sn: Jensen
+telephonenumber: +1 408 555 1212
+`
```
if I'm reading the rfc correctly, any number of spaces (including 0) are allowed between the ":" and the attribute value. would be good to add a test entry that omits the space, and one that has multiple spaces (and validate the resulting entries)

liggitt commented on ldif_test.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+var ldifBase64 = `dn: uid=someone,dc=example,dc=org
+sn:: U29tZSBPbmU=
```
maybe add a test with invalid base64

liggitt commented on ldif.go:
```diff
@@ -0,0 +1,256 @@
+			return
+		}
+		val = string(dec[:n])
+	case '<': // FIXME missing return for *net.URL type
```
should this return a `*net.URL` or inline the content located at the specified URL?

This makes it sound like an entry with `myattr:< file:///myfile.txt` should result in an entry where the `myattr` attribute is set to the content of `/myfile.txt`:

> 6) When a URL is specified in an attrval-spec, the following conventions apply:
> a) Implementations SHOULD support the file:// URL format.  The contents of the referenced file are to be included verbatim in the interpreted output of the LDIF file.
> b) Implementations MAY support other URL formats.  The semantics associated with each supported URL will be documented in an associated Applicability Statement.

liggitt commented:
thanks for the PR!

vetinari commented:
I'll fix the issues, this may take some time

vetinari commented:
pending changes for Marshal() depend on https://github.com/go-ldap/ldap/pull/54

enochtsang commented:
Hi is there something in particular keeping this PR from being merged?

johnweldon commented:
@enochtsang I think @vetinari has a couple cleanup items to finish first. 

@vetinari It looks like you've addressed all the original comments from @liggitt - what else is there to do from your perspective?

enochtsang commented:
Hi I've actually been using this branch even though is hasn't been merged yet, I've noticed that entries must be separated by exactly one line, no more. I'm not sure if this strict whitespacing is intended but the ldapadd utility is not as strict.

```
# Organization Units
dn: ou=users,dc=example,dc=com
objectClass: organizationalUnit
objectClass: top
ou: users


# searches for above empty line for dn but fails and errors out, although this is a valid ldif file for ldapadd
dn: ou=groups,dc=example,dc=com
objectClass: organizationalUnit
objectClass: top
ou: groups
```

I would like to comment that it seems like the as well as Modify not yet being supported as in the comments, Add and Delete are also not being parsed for yet as far as I can tell. if I'm not wrong, it's not a huge deal feature wise but the comments should probably add that Add and Delete are also not supported.

liggitt commented on ldif/ldif.go:
```diff
@@ -0,0 +1,327 @@
+		return nil, errors.New("empty entry?")
+	}
+
+	if l.Version == 0 && strings.HasPrefix(lines[0], "version:") {
```
seems like it would be better to make the version parsing an optional thing done on the first line of the file, rather than part of parseEntry. as it is, the first entry could be missing `version` line, and the second one have one, and this would parse it out as the ldif version

vetinari replied:
fixed now in 8232180

liggitt commented on ldif/ldif.go:
```diff
@@ -0,0 +1,327 @@
+				lines = append(lines, line)
+				entry, perr := l.parseEntry(lines)
+				if perr != nil {
+					return &ParseError{Line: curLine, Message: perr.Error()}
```
curLine here will always reference the blank line at the end of the entry, not the line in the entry that actually caused the problem

vetinari replied:
fixed with 94b60a2

vetinari commented:
@johnweldon now there's nothing left... well, except for the mod(r)dn changes, but those depend on #54

vetinari commented:
@enochtsang it's not explicitly stated that Add / Delete and Modify entries are not supported, but the head of ldif/ldif.go states that 

> This one currently just supports LDIFs like they are generated by tools like ldapsearch(1) slapcat(8). Change records are not supported while unmarshalling.

enochtsang commented on ldif/ldif_test.go:
```diff
@@ -181,4 +182,28 @@ func TestLDIFURL(t *testing.T) {
+
+
+# searches for above empty line for dn but fails and errors out in this PR
+# Even though this is a valid LDIF file for ldapadd
```
This looked a little familiar haha, these two comments aren't necessary anymore I don't think.

enochtsang commented:
Can this pull request be merged?

vetinari commented:
from my POV, yes.

johnweldon commented on ldif/apply.go:
```diff
@@ -0,0 +1,67 @@
+		case entry.Add != nil:
+			if err := conn.Add(entry.Add); err != nil {
+				if continueOnErr {
+					log.Printf("ERROR: Failed to add %s: %s", entry.Add.DN, err)
```
Maybe as a future TODO, can we use an io.Writer here, or maybe pass in the log rather than using the static global?

johnweldon commented on ldif/apply.go:
```diff
@@ -0,0 +1,67 @@
+				}
+				return fmt.Errorf("failed to modify %s: %s", entry.Modify.DN, err)
+			}
+			/*
```
Let's not check in commented code either, unless there is some good reason.

johnweldon commented on ldif/ldif.go:
```diff
@@ -0,0 +1,335 @@
+	"net/url"
+
+	"gopkg.in/ldap.v2"
+	// "os"
```
ditto

johnweldon commented on ldif/ldif.go:
```diff
@@ -0,0 +1,335 @@
+	Add    *ldap.AddRequest
+	Del    *ldap.DelRequest
+	Modify *ldap.ModifyRequest
+	//ModDN  *ldap.ModifyDNRequest
```
"

johnweldon commented on ldif/marshal.go:
```diff
@@ -0,0 +1,273 @@
+
+// Marshal returns an LDIF string from the given LDIF.
+//
+// The default line lenght is 76 characters. This can be changed by setting
```
Trivial typo `s/lenght/length/`

johnweldon commented on ldif/marshal.go:
```diff
@@ -0,0 +1,273 @@
+				data += "-\n"
+			}
+
+			/*
```
Is this block needed?  If it doesn't work, are we trying to preserve something in this commented code that requires leaving it in?  I think it's better to remove commented code.

johnweldon commented on ldif/marshal.go:
```diff
@@ -0,0 +1,273 @@
+		case *ldap.ModifyRequest:
+			l.Entries = append(l.Entries, &Entry{Modify: e.(*ldap.ModifyRequest)})
+
+			/*
```
"

johnweldon commented on ldif/marshal_test.go:
```diff
@@ -0,0 +1,272 @@
+	}
+}
+
+/*
```
"

johnweldon commented:
Other than the `ModifyDN` commented code (can we remove that or is there a compelling reason to leave that commented code in?), and a couple minor issues, this seems mergeable to me.

After addressing the trivial items and maybe filing an issue for the `io.Writer` or passed in `log` to fix in the future, I'm fine with merging this.  @liggitt any objections?

liggitt commented:
I don't object, though I haven't played with the API enough to know if there are improvements we'll want. That makes tagging versions tricky, if there is new, experimentalish code side-by-side with code we want to stay version compatible with.

liggitt commented:
I've been thinking about starting a v2 maintenance branch, and working in master to resolve some of the issues tagged for v3 that involve API changes or dropping support for old go versions. How would you feel about this just living in master until that work was done and we reached a v3.0.0 tag?

vetinari commented:
After addressing the above issues, what about putting the ldif/ dir as a new repo (https://github.com/go-ldap/ldif) and then maybe make it accessible as gopkg.in/ldif.v0 / v1?

liggitt commented:
I like the ability to version the ldif stuff separately, and starting with v0 seems like a good way to clarify the expectations around API stability until we've had more time with it

johnweldon commented:
That sounds good to me too - @liggitt do you want to make the new repo, or should I?

liggitt commented:
created at https://github.com/go-ldap/ldif

vetinari commented:
PR done: https://github.com/go-ldap/ldif/pull/3

liggitt commented:
thanks, merged
